util/fork_func.c

```diff
@@ -0,0 +1,78 @@
+#include <sys/wait.h>
+
+#include <stdint.h>
+#include <unistd.h>
+
+#include "warnp.h"
+
+#include "fork_func.h"
+
+/**
+ * fork_func(func, cookie):
+ * Fork and run ${func} in a new process, with ${cookie} as the sole argument.
+ */
+pid_t
+fork_func(int (* func)(void *), void * cookie)
+{
+	pid_t pid;
+
+	/* Fork */
+	switch (pid = fork()) {
+	case -1:
+		warnp("fork");
+		goto err0;
+	case 0:
+		/* In child process: Run the provided function, then exit. */
+		_exit((*func)(cookie));
+	default:
+		/* In parent process: do nothing else. */
+		break;
+	}
+
+	/* Success! */
+	return (pid);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/**
+ * fork_func_wait(pid):
+ * Wait for the process ${pid} to finish.  Print any error arising from ${pid}.
+ * If ${pid} exited cleanly, return its exit code; otherwise, return -1.
+ */
+int
+fork_func_wait(pid_t pid)
+{
+	int status;
+	int rc = -1;
+
+	/* Wait for the process to finish. */
+	if (waitpid(pid, &status, 0) == -1) {
+		warnp("waitpid");
+		goto err0;
+	}
+
+	/* Print the error status, if applicable. */
+	if (WIFEXITED(status)) {
+		/* Child ${pid} exited cleanly. */
+		rc = WEXITSTATUS(status);
+	} else {
+		/* Child ${pid} did not exit cleanly. */
+
+		/* Warn about the reason for the unclean exit. */
+		if (WIFSIGNALED(status))
+			warn0("pid %jd: signalled with %i",
+			    (intmax_t)pid, WTERMSIG(status));
+		else if (WIFSTOPPED(status))
```

Comment:
I figured that anything other than `WIFEXITED` was unusual and worth reporting?

I mean, the overall point of `fork_func_wait()` is "wait until the forked process is finished".  I don't have a good sense of what to do if there's an error, so I figured that reporting it was a decent lower bound?

Reply:
```
     WIFSTOPPED(status)
             True if the process has not terminated, but has stopped and can
             be restarted.  This macro can be true only if the wait call
             specified the WUNTRACED option or if the child process is being
             traced (see ptrace(2)).
```

Since we're not passing the WUNTRACED flag (or any flags at all for that matter), the only thing this would report to us is if the process is being traced... in which case the signal value isn't particularly useful anyway.  So I would just handle this with the 'unknown reason' message.

Reply:
ok, removed